Added content for session code

GloreenS opened the pull request:
### Related links

PR #284 

### Changes

Added new content for Understanding Session Code

ACStoneCL commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,174 @@
+
+4.  In the `cargo.toml` file include the following dependencies, for the purposes of this guide we are using only two, however, you can use more depending on the requirement of your session code:
+    -   `casper-contract = "1.4.3"` - You need to import the casper-contract as it provide the SDK for the execution engine (EE). For more information on this crate, see the [documentation](https://docs.rs/casper-contract/latest/casper_contract/).
+    -   `casper-types = "1.4.6"` - You need to import the casper-types crate as this crate includes the types that the node uses and the execution engine (EE) understands to interpret the session code. For more information on this crate, see the [documentation](https://docs.rs/casper-types/latest/casper_types/).  
```
```suggestion
    -   `casper-types = "1.4.6"` - You need to import the casper-types crate as this crate includes the types that the node uses. This is necessary for the execution engine (EE) to understand and interpret the session code. For more information on this crate, see the [documentation](https://docs.rs/casper-types/latest/casper_types/).  
```

ACStoneCL commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,174 @@
+-   `payment-amount` - The payment for the deploy in motes.  
+-   `session-path` - The path to the contract Wasm, which should point to wherever you compiled the contract (.wasm file) on your computer.
+
+## Comparing Session Code and Smart Contract
```
```suggestion
## Comparing Session Code and Contract Code
```
From what I understand, "contract code" is the direct comparison vs "session code", rather than using the term smart contract.

ACStoneCL commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,174 @@
+-   `session-path` - The path to the contract Wasm, which should point to wherever you compiled the contract (.wasm file) on your computer.
+
+## Comparing Session Code and Smart Contract
+The following points try to explain the difference between session code and smart contract:
```
```suggestion
The following points try to explain the difference between session code and contract code.
```

ACStoneCL commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,174 @@
+
+## Comparing Session Code and Smart Contract
+The following points try to explain the difference between session code and smart contract:
+- When `put_key` is used to store a URef, it depends where the key is stored based on if the session code or the smart contract is making the call. When the session code is executed the key is added to the named keys of the account that called the session code. Whereas, when a smart contract is executed the key is added to the named keys of the smart contract. 
```
```suggestion
- When `put_key` is used to store a URef, it depends where the key is stored based on if the session code or the contract code is making the call. When the session code is executed the key is added to the named keys of the account that called the session code. Whereas, when contract code is executed the key is added to the named keys of the smart contract. 
```

ACStoneCL commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,174 @@
+## Comparing Session Code and Smart Contract
+The following points try to explain the difference between session code and smart contract:
+- When `put_key` is used to store a URef, it depends where the key is stored based on if the session code or the smart contract is making the call. When the session code is executed the key is added to the named keys of the account that called the session code. Whereas, when a smart contract is executed the key is added to the named keys of the smart contract. 
+- Smart contract is Wasm that is stored on the network. Whereas session code is simple code you can use to interact with the smart contract.
```
```suggestion
- Contract code is Wasm that is stored on the network. Whereas session code is simple code you can use to interact with the smart contract.
```

ACStoneCL replied:
```suggestion
- A smart contract is Wasm that is stored on the network. Whereas session code is simple code you can use to interact with the smart contract.
```

ACStoneCL commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,174 @@
+The following points try to explain the difference between session code and smart contract:
+- When `put_key` is used to store a URef, it depends where the key is stored based on if the session code or the smart contract is making the call. When the session code is executed the key is added to the named keys of the account that called the session code. Whereas, when a smart contract is executed the key is added to the named keys of the smart contract. 
+- Smart contract is Wasm that is stored on the network. Whereas session code is simple code you can use to interact with the smart contract.
+- Smart contract is stateful and session code is stateless.
```
```suggestion
- A smart contract is stateful and session code is stateless.
```

ACStoneCL reviewed (commented):
I've added some flow suggestions, and changed some instances of 'smart contract' to 'contract code' in other suggestions.

I was also using the 'Adder' smart contract as an example in my return values tutorial. When I reviewed it with @EdHastingsCasperLabs , he said that he wanted to avoid using a contract that performs a function that the platform can do - as to avoid anyone assuming that they need to program their own basic math functions if they decide to use Casper.

@darthsiroftardis created a basic 'donation' contract that can be easily switched out for the example code portion.

Other than that, the content looks good.

ipopescu reviewed (changes requested):
Please rebase against the dev branch.

GloreenS commented:
> I've added some flow suggestions, and changed some instances of 'smart contract' to 'contract code' in other suggestions.
> 
> I was also using the 'Adder' smart contract as an example in my return values tutorial. When I reviewed it with @EdHastingsCasperLabs , he said that he wanted to avoid using a contract that performs a function that the platform can do - as to avoid anyone assuming that they need to program their own basic math functions if they decide to use Casper.
> 
> @darthsiroftardis created a basic 'donation' contract that can be easily switched out for the example code portion.
> 
> Other than that, the content looks good.

Thanks @ACStoneCL, I have updated the session code example with "donation_count" and added your other suggestions as well.

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+
+## What is Session code?
+Session code is the simplest piece of logic you can execute on a Casper network. It requires only one entry point, which is the `call` function and it runs within the context of the account executing the session code. This means that the session code will run with the permissions of the account, such as having access to the main purse (the session code could transfer tokens out of the account's m ...
+Session code can be written in any programming language that compiles to Wasm.
```
```suggestion
Session code can be written in any programming language that compiles to Wasm. But, in the examples below we will use Rust.
```

ipopescu replied:
```suggestion
Session code can be written in any programming language that compiles to Wasm. But, in the examples below we will use Rust.
```
```suggestion
Session code can be written in any programming language that compiles to Wasm.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+    |_ Cargo.lock
+    |_ Cargo.toml
+
+In the above directory structure the `contract` folder contains the session code in the `main.rs` file and the needed dependencies in the `Cargo.toml` file. 
+The `tests` folder contains the code required to test the session code on before it is deployed on a Casper network.
```
```suggestion
In the above directory structure, the `contract` folder contains the session code in the `main.rs` file and the needed dependencies in the `Cargo.toml` file. 
The `tests` folder contains the code required to test the session code before it is deployed on a Casper network.
```

ipopescu replied:
```suggestion
In the above directory structure, the `contract` folder contains the session code in the `main.rs` file and the needed dependencies in the `Cargo.toml` file. 
The `tests` folder contains the code required to test the session code before it is deployed on a Casper network.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+The `tests` folder contains the code required to test the session code on before it is deployed on a Casper network.
+
+## Writing Session Code
+The following steps illustrate the process of writing session code and the important components to include:
```
```suggestion
The following steps illustrate the process of writing session code and the important components to include.
```
```suggestion
The following steps illustrate the process of writing session code and the important components to include:
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+The `tests` folder contains the code required to test the session code on before it is deployed on a Casper network.
+
+## Writing Session Code
+The following steps illustrate the process of writing session code and the important components to include:
```
```suggestion
The following steps illustrate the process of writing session code and the important components to include.
```

GloreenS replied:
Colon is the correct usage here.

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+
+3. Within the contract package, you can find the `main.rs` file inside the `src` folder. You will write your session code in the `main.rs` file. 
+
+4.  In the `cargo.toml` file include the following dependencies, for the purposes of this guide we are using only two, however, you can use more depending on the requirement of your session code:
```
Split the sentence here.
```suggestion
4.  In the `cargo.toml` file, include the following dependencies. For the purposes of this guide, we are using only two dependencies; however, you can use more depending on the requirement of your session code:
```

ipopescu replied:
Split the sentence here.
```suggestion
4.  In the `cargo.toml` file, include the following dependencies. For the purposes of this guide, we are using only two dependencies; however, you can use more depending on the requirement of your session code:
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+3. Within the contract package, you can find the `main.rs` file inside the `src` folder. You will write your session code in the `main.rs` file. 
+
+4.  In the `cargo.toml` file include the following dependencies, for the purposes of this guide we are using only two, however, you can use more depending on the requirement of your session code:
+    -   `casper-contract = "1.4.3"` - You need to import the casper-contract as it provide the SDK for the execution engine (EE). For more information on this crate, see the [documentation](https://docs.rs/casper-contract/latest/casper_contract/).
```
I would put the link with the first mention of casper-contract.
```suggestion
    -   `casper-contract = "1.4.3"` - You need to import the [casper-contract](https://docs.rs/casper-contract/latest/casper_contract/) as it provides the SDK for the execution engine (EE).
```

ipopescu replied:
```suggestion
    -   `casper-contract = "1.4.3"` - You need to import the [casper-contract](https://docs.rs/casper-contract/latest/casper_contract/) as it provides the SDK for the execution engine (EE).
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+
+4.  In the `cargo.toml` file include the following dependencies, for the purposes of this guide we are using only two, however, you can use more depending on the requirement of your session code:
+    -   `casper-contract = "1.4.3"` - You need to import the casper-contract as it provide the SDK for the execution engine (EE). For more information on this crate, see the [documentation](https://docs.rs/casper-contract/latest/casper_contract/).
+    -   `casper-types = "1.4.6"` - You need to import the casper-types crate as this crate includes the types that the node uses. This is necessary for the execution engine (EE) to understand and interpret the session code. For more information on this crate, see the [documentation](https://docs.rs/casper-types/latest/casper_types/).  
```
```suggestion
    -   `casper-types = "1.4.6"` - You need to import the [casper-types](https://docs.rs/casper-types/latest/casper_types/) crate as this crate includes the types that the node uses. This is necessary for the execution engine (EE) to understand and interpret the session code.
```

ipopescu replied:
```suggestion
    -   `casper-types = "1.4.6"` - You need to import the [casper-types](https://docs.rs/casper-types/latest/casper_types/) crate as this crate includes the types that the node uses. This is necessary for the execution engine (EE) to understand and interpret the session code.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+
+    You can find the latest versions of the dependencies at https://crates.io/.
+    
+5. Few things to note while writing session code:
```
```suggestion
5. Note the following suggestions while writing session code.
```

ipopescu replied:
```suggestion
5. A few things to note while writing session code:
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+    
+5. Few things to note while writing session code:
+    -   Include the following:
+        -   `#![no_std]` - This indicates to not import the standard library.
```
```suggestion
        -   `#![no_std]` - This indicates not to import the standard library.
```

ipopescu replied:
```suggestion
        -   `#![no_std]` - This indicates not to import the standard library.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+5. Few things to note while writing session code:
+    -   Include the following:
+        -   `#![no_std]` - This indicates to not import the standard library.
+        -   `#![no_main]` - This indicates that the main function is not required, since the session code has only one entry point as the `call` function.
```
Should we highlight the main function as you did the `call` function?

ipopescu replied:
Should we highlight the main function as you did the `call` function?

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+        -   `#![no_std]` - This indicates to not import the standard library.
+        -   `#![no_main]` - This indicates that the main function is not required, since the session code has only one entry point as the `call` function.
+    -   Import the casper contract API:
+        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses account, runtime, storage, and system crates, however, you might need to import the crates relevant to your session code.
```
In general, we use semicolons before however.
```suggestion
        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses the account, runtime, storage, and system crates; however, you might need to import the crates relevant to your session code.
```

GloreenS replied:
That is an incorrect assumption, there is no such grammatical rule. Semicolon is used to join two related independent clauses in place of a comma and a coordinating conjunction. 
"However" doesn't fall into this rule always!

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+        -   `#![no_std]` - This indicates to not import the standard library.
+        -   `#![no_main]` - This indicates that the main function is not required, since the session code has only one entry point as the `call` function.
+    -   Import the casper contract API:
+        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses account, runtime, storage, and system crates, however, you might need to import the crates relevant to your session code.
```
```suggestion
        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses the account, runtime, storage, and system crates; however, you might need to import the crates relevant to your session code.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses account, runtime, storage, and system crates, however, you might need to import the crates relevant to your session code.
+
+## Sample Session Code
+This sample code demonstrates a simple session code passing arguments, processing the arguments, and storing the result. In general, that is what you will use session code for. In this example, we are adding two numbers and storing the result in a URef.
```
Adder is not used anymore, updated explanation needed.
```suggestion
This sample code demonstrates a simple session code passing arguments, processing the arguments, and storing the result. In general, you will use session code for such operations. In this example, we call a contract that gets a donation amount, and we store the result in a URef.
```

ipopescu replied:
This is an important one.

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses account, runtime, storage, and system crates, however, you might need to import the crates relevant to your session code.
+
+## Sample Session Code
+This sample code demonstrates a simple session code passing arguments, processing the arguments, and storing the result. In general, that is what you will use session code for. In this example, we are adding two numbers and storing the result in a URef.
```
```suggestion
This sample code demonstrates a simple session code passing arguments, processing the arguments, and storing the result. In general, you will use session code for such operations. In this example, we add two numbers and store the result in a URef.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+#![no_std]
+#![no_main]
+```
+This indicates to not import the standard library and that the main function is not required, since the session code has only one entry point as the `call` function.
````
Please move all explanations before the code. Please do so for the rest of the code.
```suggestion
This indicates not to import the standard library and that the main function is not required, since the session code has only one entry point as the `call` function.
```

ipopescu replied:
```suggestion
This indicates not to import the standard library and that the main function is not required, since the session code has only one entry point as the `call` function.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+use casper_contract::unwrap_or_revert::UnwrapOrRevert;
+use casper_types::{runtime_args, ContractHash, Key, PublicKey, RuntimeArgs, URef, U512};
+```
+Imports the casper contract API. This example uses account, runtime, storage, and system crates, however, you might need to import the crates relevant to your session code.
````
Generally, I see semicolons before however.

ipopescu replied:
Generally, I see semicolons before however.
```suggestion
Imports the casper contract API. This example uses the account, runtime, storage, and system crates; however, you might need to import the crates relevant to your session code.
```

ipopescu replied:
Generally, I see semicolons before however.
```suggestion
Imports the casper contract API. This example uses the account, runtime, storage, and system crates; however, you might need to import the crates relevant to your session code.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+```rust
+#[no_mangle]
+```
+When some Wasm is sent to be executed by the execution engine (EE) that lives within each node of a Casper network, `#[no_mangle]` ensures that the function name that follows it is retained as is in string form in the Wasm output. For session code, this retains the `call` string and marks the entry point for the execution engine.
````
Hard to read sentence. Suggesting a rewrite. Also, if you indicate the acronym (EE), it's best to indicate it only once at the start of the file, and start using it. I see it indicated several times.
Output doesn't seem to fit, I would replace it with bytecode unless the EE outputs the Wasm again? Probably best to check with @darthsiroftardis here.
```suggestion
When the EE (that lives on each node of a Casper network), receives Wasm to execute, the `#[no_mangle]` flag ensures that the function name following it is retained as a string in the Wasm output. For session code, this retains the `call` string and marks the entry point for the execution engine.
```

ipopescu replied:
Helpful if implemented.

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+```rust
+#[no_mangle]
+```
+When some Wasm is sent to be executed by the execution engine (EE) that lives within each node of a Casper network, `#[no_mangle]` ensures that the function name that follows it is retained as is in string form in the Wasm output. For session code, this retains the `call` string and marks the entry point for the execution engine.
````
Hard to read sentence. Suggesting a rewrite. Also, if you indicate the acronym (EE), it's best to indicate it only once at the start of the file, and start using it. I see it indicated several times.
Output doesn't seem to fit, I would replace it with bytecode unless the EE outputs the Wasm again? Probably best to check with @darthsiroftardis here.
```suggestion
When the EE (that lives on each node of a Casper network), receives Wasm to execute, the `#[no_mangle]` flag ensures that the function name following it is retained as a string in the Wasm output. For session code, this retains the `call` string and marks the entry point for the execution engine.
```

ipopescu replied:
Discussed with Gloreen.

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+```rust
+pub extern "C" fn call()
+```
+This initiates the `call` function. When compiled, the resulting Wasm could then be linked to from a C library, and the function could be used as if it was from any other library.
````
Hard to read. Suggesting a rewrite.
```suggestion
This initiates the `call` function, which when compiled could be used from another library. For example, a C library could link to the resulting Wasm.
```

ipopescu replied:
Discussed with Gloreen.

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+```rust
+pub extern "C" fn call()
+```
+This initiates the `call` function. When compiled, the resulting Wasm could then be linked to from a C library, and the function could be used as if it was from any other library.
````
Hard to read. Suggesting a rewrite.
```suggestion
This initiates the `call` function, which when compiled could be used from another library. For example, a C library could link to the resulting Wasm.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+        },
+    );
+```
+This piece of code tries to demonstrate how to get the contract hash and donating account key as arguments and then perform a simple operation with them, such as get the final donation count. The `runtime::get_named_arg()` takes a string as an argument and returns the named argument to the host in the current runtime.
````
Hard to read. Suggestion to break the sentence down.
```suggestion
This code demonstrates how to get the contract hash and donating account key as arguments. It then performs a simple operation with them, such as getting the final donation count. The `runtime::get_named_arg()` takes a string as an argument and returns the named argument to the host in the current runtime.
```

ipopescu replied:
Hard to read. Suggestion to break the sentence down.
```suggestion
This code demonstrates how to get the contract hash and donating account key as arguments. It then performs a simple operation with them, such as getting the final donation count. The `runtime::get_named_arg()` takes a string as an argument and returns the named argument to the host in the current runtime.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+```rust
+let donation_count_uref = storage::new_uref(donation_count);
+```
+Once you have the result, you might want to save it at a location that can be accessed later. This code puts the URef in the current context's NamedKeys, which is the context of the account calling this piece of session code.
````
Suggestion to explain NamedKeys or to link somewhere that explains them. At a minimum here https://docs.rs/casper-types/latest/casper_types/contracts/type.NamedKeys.html.

ipopescu replied:
Suggestion to explain NamedKeys or to link somewhere that explains them. At a minimum here https://docs.rs/casper-types/latest/casper_types/contracts/type.NamedKeys.html.

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+```rust    
+runtime::put_key("donation_count", donation_count_uref.into())
+```
+The `put_key` function stores the URef of the result in the current context's NamedKeys, which is the context of the account calling this piece of session code. Once this session code is executed, the account that called the session code will have a new named key `donation_count` added to the account.
````
Please rewrite this sentence as it is hard to read. I have done this for some of the examples above.

ipopescu replied:
Please rewrite this sentence as it is hard to read. I have done this for some of the examples above.

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
````diff
@@ -0,0 +1,190 @@
+```bash
+cd contract
+```
+Inside the *contract* directory execute the following command to compile the session code.
````
```suggestion
Inside the *contract* directory, execute the following command to compile the session code.
```
```suggestion
Inside the *contract* directory execute the following command to compile the session code.
```

ipopescu replied:
```suggestion
Inside the *contract* directory, execute the following command to compile the session code.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+Once the session code is compiled you can deploy it on a Casper network.
+
+## Deploying the Session Code
+Before you deploy the session code to the Mainnet or Testnet, you can do a trial run on the a local network using NCTL. For more information on how to build an NCTL network, see [Local Network Testing](setup-nctl.md).
```
Best to link directly.
```suggestion
Before you deploy the session code to the Mainnet or Testnet, you can do a trial run on a local network using [NCTL](setup-nctl.md).
```

ipopescu replied:
Best to link directly. Also, note the extra article.
```suggestion
Before you deploy the session code to the Mainnet or Testnet, you can do a trial run on a local network using [NCTL](setup-nctl.md).
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+-   `payment-amount` - The payment for the deploy in motes.  
+-   `session-path` - The path to the contract Wasm, which should point to wherever you compiled the contract (.wasm file) on your computer.
+
+## Comparing Session Code and Contract Code
```
I think this can be a sub-point in the introduction, with links to the appropriate sub-sections or pages (such as writing contracts).

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+-   `payment-amount` - The payment for the deploy in motes.  
+-   `session-path` - The path to the contract Wasm, which should point to wherever you compiled the contract (.wasm file) on your computer.
+
+## Comparing Session Code and Contract Code
```
I think this can be a sub-point in the introduction, with links to the appropriate sub-sections or pages (such as writing contracts).

GloreenS replied:
Done, moved.

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+
+## Comparing Session Code and Contract Code
+The following points try to explain the difference between session code and contract code.
+- When `put_key` is used to store a URef, it depends where the key is stored based on if the session code or the contract code is making the call. When the session code is executed the key is added to the named keys of the account that called the session code. Whereas, when contract code is executed the key is added to the named keys of the smart contract. 
```
Please rewrite this, as it is hard to read. Also, you can link to the example on line 148 if you wish. 
```suggestion
- When `put_key` is used to store a URef, the key is stored differently depending on if the caller is session code or contract code. When session code is executed, the key is added to the named keys of the account calling the session code. When contract code is executed, the key is added to the named keys of the smart contract. 
```

ipopescu replied:
Please rewrite this, as it is hard to read. Also, you can link to the example on line 148 if you wish. 
```suggestion
- When `put_key` is used to store a URef, the key is stored differently depending on if the caller is session code or contract code. When session code is executed, the key is added to the named keys of the account calling the session code. When contract code is executed, the key is added to the named keys of the smart contract. 
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code.md:
```diff
@@ -0,0 +1,190 @@
+## Comparing Session Code and Contract Code
+The following points try to explain the difference between session code and contract code.
+- When `put_key` is used to store a URef, it depends where the key is stored based on if the session code or the contract code is making the call. When the session code is executed the key is added to the named keys of the account that called the session code. Whereas, when contract code is executed the key is added to the named keys of the smart contract. 
+- Contract code is Wasm that is stored on the network. Whereas, session code is simple code you can use to interact with the smart contract.
```
```suggestion
- Contract code is Wasm that is stored on the network. Whereas, session code is simpler code used to interact with contract code.
```

ipopescu replied:
```suggestion
- Contract code is Wasm that is stored on the network. Whereas, session code is simpler code used to interact with contract code.
```

ipopescu reviewed (commented):
This is a very good article, @GloreenS. I think it is technically sound.
I have left several minor suggestions to address before the merge.
If possible, please move all explanations before the code, and also, consider rewriting a couple of hard-to-read sentences (I left comments where I thought this was the case). Many thanks for the article, it is great!

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,197 @@
+# Writing Session Code
+This section explains the concept of session code, why we need it, and how to write it. The best use of session code is when the situation calls for a [stateless](../../glossary/S.md/#stateless) execution. Session code is useful when little or no internal state is required. You can use session code when the logic requires very little or no internal data to be tracked, or to simplify the process o ...
```
```suggestion
This section explains the concept of session code, why we need it, and how to write it. The best use of session code is when the situation calls for a [stateless](../../glossary/S.md/#stateless) execution. Session code is useful when little or no internal state is required. You can use session code when the logic requires very little or no internal data to be tracked, or to simplify the process of interacting with entry points of a contract.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,197 @@
+- Session code and contract code run in two different type of contexts. Session code always executes in the context of the account that signed the deploy that contains the session code. This means that when a `put_key` call is made within the body of the session code, the key is added to the account's named keys. 
+- Conversely, contract code executes in its own context. Which means that when `put-key` call is made within the contract's execution, the key is inserted into the contract's context. So, the key will appear in the contract's named keys.
+- Session code has only one entry point, that is the `call` function, which you can use to interact with the session code. 
+- A contract code can have multiple entry points that will help you interact with the contract code. 
```
I recommend saying "A contract.." or "contract code...". 
```suggestion
- A contract can have multiple entry points that will help you interact with the contract code. 
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,197 @@
+## Writing Session Code
+The following steps illustrate the process of writing session code and the important components to include:
+
+1. Create a new top level directory that will contain the session code and would also include another folder for tests, which will help us test the functionality of our session code. 
```
```suggestion
1. Create a new top-level directory containing the session code and include another folder for tests, which will help us test the functionality of our session code. 
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,197 @@
+
+1. Create a new top level directory that will contain the session code and would also include another folder for tests, which will help us test the functionality of our session code. 
+
+2. Inside the new folder run the following command to create a new binary package called contract:
```
```suggestion
2. Inside the new folder run the following command to create a new binary package called "contract":
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,197 @@
+
+3. Within the contract package, you can find the `main.rs` file inside the `src` folder. You will write your session code in the `main.rs` file. 
+
+4.  In the `cargo.toml` file include the following dependencies, for the purposes of this guide we are using only two, however, you can use more depending on the requirement of your session code:
```
The sentence is too long.
```suggestion
4.  In the `cargo.toml` file include the following dependencies. For the purposes of this guide, we are using only two, however, you can use more depending on the requirement of your session code:
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,197 @@
+
+    You can find the latest versions of the dependencies at https://crates.io/.
+    
+5. Few things to note while writing session code:
```
Please add the article.
```suggestion
5. A few things to note while writing session code:
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,197 @@
+        -   `#![no_std]` - This indicates to not import the standard library.
+        -   `#![no_main]` - This indicates that the main function is not required, since the session code has only one entry point as the `call` function.
+    -   Import the casper contract API:
+        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses account, runtime, storage, and system crates, however, you might need to import the crates relevant to your session code.
```
```suggestion
        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses account, runtime, storage, and system crates; however, you might need to import the crates relevant to your session code.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,197 @@
+        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses account, runtime, storage, and system crates, however, you might need to import the crates relevant to your session code.
+
+## Sample Session Code
+This sample code demonstrates a simple session code passing arguments, processing the arguments, and storing the result. In general, that is what you will use session code for. In this example, we are adding two numbers and storing the result in a URef.
```
Try avoid ending with "for"
```suggestion
This sample code demonstrates a simple session code passing arguments, processing the arguments, and storing the result. In general, for this, you will use session code. In this example, we add two numbers and store the result in a URef.
```

ipopescu commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
````diff
@@ -0,0 +1,197 @@
+### Understanding the sample code
+Let's try to understand what each line of code in the above sample is trying to achieve.
+
+```rust
````
I really think adding the code after the explanation is much better -- unless there is a common pattern/recommendation you are trying to follow?

ipopescu replied:
Discussed with Gloreen, it is better the way it is.

ipopescu reviewed (commented):
@GloreenS, please take another look at my previous suggestions; many of them did not make it to the latest version, although I took the time to run the content through Grammarly and placed them in an easy "commit" format. Many sentences can be shorter or simplified so that the reader can focus on the technical information. I have stopped my second review at line 109.
When you are ready, let me know, and I will do a final review to unblock the PR.

ipopescu commented:
I understand, but I mentioned it not because of grammar rules, but because
I find the sentence long and technical, thus hard to keep track of all the
pieces.


On Fri, May 6, 2022 at 9:00 AM GloreenS ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In source/docs/casper/dapp-dev-guide/session-code.md
> <https://github.com/casper-network/docs/pull/315#discussion_r866541278>:
>
> > +    This folder will contain the logic that will be compiled to Wasm and will be executed on a node within a Casper network.
> +
> +3. Within the contract package, you can find the `main.rs` file inside the `src` folder. You will write your session code in the `main.rs` file.
> +
> +4.  In the `cargo.toml` file include the following dependencies, for the purposes of this guide we are using only two, however, you can use more depending on the requirement of your session code:
> +    -   `casper-contract = "1.4.3"` - You need to import the casper-contract as it provide the SDK for the execution engine (EE). For more information on this crate, see the [documentation](https://docs.rs/casper-contract/latest/casper_contract/).
> +    -   `casper-types = "1.4.6"` - You need to import the casper-types crate as this crate includes the types that the node uses. This is necessary for the execution engine (EE) to understand and interpret the session code. For more information on this crate, see the [documentation](https://docs.rs/casper-types/latest/casper_types/).
> +
> +    You can find the latest versions of the dependencies at https://crates.io/.
> +
> +5. Few things to note while writing session code:
> +    -   Include the following:
> +        -   `#![no_std]` - This indicates to not import the standard library.
> +        -   `#![no_main]` - This indicates that the main function is not required, since the session code has only one entry point as the `call` function.
> +    -   Import the casper contract API:
> +        `use casper_contract::contract_api::{account, runtime, storage, system};` this example uses account, runtime, storage, and system crates, however, you might need to import the crates relevant to your session code.
>
> That is an incorrect assumption, there is no such grammatical rule.
> Semicolon is used to join two related independent clauses in place of a
> comma and a coordinating conjunction.
> "However" doesn't fall into this rule always!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/casper-network/docs/pull/315#discussion_r866541278>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA757CRQ3FFGSDQGRFAXPCDVIS7PPANCNFSM5SETNDGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ipopescu commented on config/sidebar.config.js:
```diff
@@ -40,6 +40,28 @@ module.exports = {
+            label: "Writing Session Code",
+            collapsible: true,
+            collapsed: true,
+            items: ["dapp-dev-guide/session-code/index", "dapp-dev-guide/session-code/session-code"],
```
Testing will come as well.

ipopescu reviewed (approved):
Discussed all changes with Gloreen manually and everything looks good now.

GloreenS commented:
> There is some changes that need to be made, suggestions from @ipopescu should also be addressed and/or applied before a second pass at the review

@darthsiroftardis the changes are now implemented.

darthsiroftardis commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,204 @@
+# Writing Session Code
+This section explains the concept of session code, why we need it, and how to write it. The best use of session code is when the situation calls for a [stateless](../../glossary/S.md/#stateless) execution. Session code is useful when little or no internal state is required. You can use session code when the logic requires very little or no internal data to be tracked, or to simplify the process o ...
```
```suggestion
This section explains the concept of session code, why we need it, and how to write it. The best use of session code is when the situation calls for a [stateless](../../glossary/S.md/#stateless) execution. You can use session code when the logic requires very little or no internal data to be tracked. Session code is required when interacting and accepting values being returned across the Wasm boundary
```

darthsiroftardis commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,204 @@
+-   `chain-name` - The chain-name to the network where you wish to send the deploy. For Mainnet, use *casper*. For Testnet, use *casper-test*. 
+-   `payment-amount` - The payment for the deploy in motes.  
+-   `session-path` - The path to the contract Wasm, which should point to wherever you compiled the contract (.wasm file) on your computer.
+
```
Since the session code we are showcasing here takes arguments, we should should make a call out to the `session-arg` field within the subcommand. Additionally, a call out to using the help for the `put-deploy` command which gives examples on passing runtime arguments using the casper command line client

GloreenS replied:
Added the details.

darthsiroftardis commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,204 @@
+The `tests` folder contains the code required to test the session code before it is deployed on a Casper Network.
+
+## Writing Session Code
+The following steps illustrate the process of writing session code and the important components to include:
```
Lets add a paragraph explaining our high level goals, i.e we are going to demonstrate how to pass runtime arguments to Wasm and then use those runtime arguments to call into another piece of logic. Lets provide some overall narrative context here

GloreenS replied:
@darthsiroftardis we have a similar description in the following section containing the sample code. In this particular section, we are not doing any of those things, but just identifying the steps in writing session code. I have changed the title of this section to "Process of Writing Session Code", since it was clashing with the main title of the topic.

darthsiroftardis reviewed (commented):
The content looks good, I'd like to see some high level narrative context, i.e what are we doing, and what you will see at the end of the section

darthsiroftardis commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,208 @@
+The following points try to explain the difference between session code and contract code:
+
+- Session code and contract code run in two different type of contexts. Session code always executes in the context of the account that signed the deploy that contains the session code. This means that when a `put_key` call is made within the body of the session code, the key is added to the account's named keys. 
+- Conversely, contract code executes in its own context. Which means that when `put-key` call is made within the contract's execution, the key is inserted into the contract's context. So, the key will appear in the contract's named keys.
```
```suggestion
- Conversely, contract code executes in its own context. Which means that when `put_key` call is made within the contract's execution, the key is inserted into the contract's context. So, the key will appear in the contract's named keys.
```

darthsiroftardis commented on source/docs/casper/dapp-dev-guide/session-code/session-code.md:
```diff
@@ -0,0 +1,208 @@
+    -   `casper-contract = "1.4.3"` - You need to import the [casper-contract](https://docs.rs/casper-contract/latest/casper_contract/) as it provides the SDK for the execution engine (EE).
+    -   `casper-types = "1.4.6"` - You need to import the [casper-types](https://docs.rs/casper-types/latest/casper_types/) crate as this crate includes the types that the node uses. This is necessary for the execution engine (EE) to understand and interpret the session code.  
+
+    You can find the latest versions of the dependencies at https://crates.io/.
```
@GloreenS Is it possible to link to the crates themselves

GloreenS replied:
Done!

darthsiroftardis reviewed (commented):
Last two comments and then I can we close this one out

ipopescu commented:
@darthsiroftardis @GloreenS I'd like to see this doc live before Monday's conference. If the checks pass, I will merge it.

ipopescu commented:
@GloreenS please try to resolve the build and merge on Monday if possible.

> These sidebar document ids do not exist:
> - dapp-dev-guide/session-code/index
> - dapp-dev-guide/session-code/session-code
